### PR TITLE
fix for whitelisting "about:blank"

### PR DESCRIPTION
**Pull request by heidijinxujia**
The "about:blank" should also be considered in the redirect URL loading checking.
see details #585

### CODE REVIEW

File: src/src/com/microsoft/aad/adal/BasicWebViewClient.java [hunk 1]
```diff
@@ -269,6 +269,9 @@ public void run() {
             openLinkInBrowser(parameters.get(ApplicationReceiver.INSTALL_URL_KEY));
             view.stopLoading();
             return true;
+        } else if (url.startsWith("about:blank")) {
```
Comment (weijjia) [inline src/src/com/microsoft/aad/adal/BasicWebViewClient.java]: IMO, this should be moved to the place where you check for SSL

File: src/src/com/microsoft/aad/adal/AuthenticationActivity.java [hunk 2]
```diff
@@ -627,18 +626,17 @@ public boolean processInvalidUrl(final WebView view, String url) {
...
                 return true;
-            }
-            
-            //check if the redirect URL is under SSL protected
-            if(!url.toLowerCase(Locale.US).startsWith(AuthenticationConstants.Broker.REDIRECT_SSL_PREFIX))
-            {
+            } else if (url.startsWith("about:blank")) {
+                Logger.v(TAG, "It is an blank page request");
+                return true;
+            } else if(!url.toLowerCase(Locale.US).startsWith(AuthenticationConstants.Broker.REDIRECT_SSL_PREFIX)){
```
Comment (weijjia) [inline src/src/com/microsoft/aad/adal/AuthenticationActivity.java]: I would rather have a new method i.e.
boolean isValidRedicectUrl(final String url) {
    return url.toLowerCase(Locale.US).startsWith(AuthenticationConstants.Broker.REDIRECT_SSL_PREFIX) || url.equalsIgnoreCase("about:blank");
}

File: src/src/com/microsoft/aad/adal/AuthenticationActivity.java [hunk 3]
```diff
@@ -627,18 +626,17 @@ public boolean processInvalidUrl(final WebView view, String url) {
                         mRedirectUrl));
                 view.stopLoading();
                 return true;
-            }
-            
-            //check if the redirect URL is under SSL protected
-            if(!url.toLowerCase(Locale.US).startsWith(AuthenticationConstants.Broker.REDIRECT_SSL_PREFIX))
-            {
+            } else if (url.startsWith("about:blank")) {
```
Comment (weijjia) [inline src/src/com/microsoft/aad/adal/AuthenticationActivity.java]: shouldn't we compare if it's equals to?

File: src/src/com/microsoft/aad/adal/AuthenticationActivity.java (hunk 2 above)
Comment (heidijinxujia) [inline src/src/com/microsoft/aad/adal/AuthenticationActivity.java]: @weijjia  In this function, you treated SSL-protected page and "about:blank" as the same, where as a result, not only the ssl-protected page will be opened in the webview, but also the "about:blank" empty page. I am wondering in which situation we need to open an blank webpage. Should we override the "about:blank" url loading, or not? Thanks.

File: src/src/com/microsoft/aad/adal/AuthenticationActivity.java (hunk 3 above)
Comment (heidijinxujia) [inline src/src/com/microsoft/aad/adal/AuthenticationActivity.java]: Fixed in the latest commit

File: src/src/com/microsoft/aad/adal/AuthenticationActivity.java (hunk 2 above)
Comment (weijjia) [inline src/src/com/microsoft/aad/adal/AuthenticationActivity.java]: IMO, you're checking if the redirect is invalid or not, and with your fix, you treat about:blank as a valid one, then both ssl protected page and about:blank are valid.

Comment (heidijinxujia) [inline src/src/com/microsoft/aad/adal/AuthenticationActivity.java]: The "about:blank" is treated as invalid redirect URL here. The function processInvalidUrl returns true if the URL is invalid. 
....
else if (url.startsWith("about:blank")) {  
                Logger.v(TAG, "It is an blank page request");  
               return true;  
} ....

File: src/src/com/microsoft/aad/adal/AuthenticationActivity.java [hunk 4]
```diff
@@ -627,18 +626,17 @@ public boolean processInvalidUrl(final WebView view, String url) {
...
+            } else if (url.toLowerCase(Locale.US).equals("about:blank")) {
+                Logger.v(TAG, "It is an blank page request");
+                return true;
+            } else if(!url.toLowerCase(Locale.US).startsWith(AuthenticationConstants.Broker.REDIRECT_SSL_PREFIX)){
             	Logger.e(TAG + methodName, "The webview was redirected to an unsafe URL.", "", ADALError.WEBVIEW_REDIRECTURL_NOT_SSL_PROTECTED);
             	returnError(ADALError.WEBVIEW_REDIRECTURL_NOT_SSL_PROTECTED, "The webview was redirected to an unsafe URL.");
             	view.stopLoading();
             	return true;
+            } else {
+            	return false;
```
Comment (RandalliLama) [inline src/src/com/microsoft/aad/adal/AuthenticationActivity.java]: This needs to be indented.

File: src/src/com/microsoft/aad/adal/AuthenticationActivity.java [hunk 5]
```diff
@@ -627,18 +626,17 @@ public boolean processInvalidUrl(final WebView view, String url) {
...
                 return true;
-            }
-            
-            //check if the redirect URL is under SSL protected
-            if(!url.toLowerCase(Locale.US).startsWith(AuthenticationConstants.Broker.REDIRECT_SSL_PREFIX))
-            {
+            } else if (url.toLowerCase(Locale.US).equals("about:blank")) {
+                Logger.v(TAG, "It is an blank page request");
+                return true;
+            } else if(!url.toLowerCase(Locale.US).startsWith(AuthenticationConstants.Broker.REDIRECT_SSL_PREFIX)){
```
Comment (addev-ashish) [inline src/src/com/microsoft/aad/adal/AuthenticationActivity.java]: nit: 2 spaces missing one after if and once before {
} else if (textTextText) {

File: src/src/com/microsoft/aad/adal/AuthenticationActivity.java (hunk 2 above)
Comment (heidijinxujia) [inline src/src/com/microsoft/aad/adal/AuthenticationActivity.java]: @weijjia

Comment (weijjia) [inline src/src/com/microsoft/aad/adal/AuthenticationActivity.java]: I'm fine with your current change. Make sure you apply all the comments.

File: tests/Functional/src/com/microsoft/aad/adal/test/AuthenticationActivityUnitTest.java [hunk 6]
```diff
@@ -279,6 +279,25 @@ public void testWebview_sslprotectedredirectURL() throws IllegalArgumentExceptio
...
+        startActivity(intentToStartActivity, null, null);
+        activity = getActivity();
+        /*
+         * case 1: url = "about:blank"
+         */
+        String url = "about:blank";
+        WebViewClient client = getCustomWebViewClient();
+        WebView mockview = new WebView(getActivity().getApplicationContext());
+        ReflectionUtils.setFieldValue(activity, "mSpinner", null);
+        //shouldOverrideUrlLoading should prevent non https redirects in the web view
```
Comment (weijjia) [inline tests/Functional/src/com/microsoft/aad/adal/test/AuthenticationActivityUnitTest.java]: remove code that is commented out.

**Comment by weijjia** [comment]
sign off after you apply all the comments.

**Comment by heidijinxujia** [comment]
Applied all the comments and plan to sign off.

**Comment by weijjia** [comment]
:shipit:

**Pull request merged**